### DISCUSSION TITLE
Catching routes based on path parameter types

### DISCUSSION THREAD

**Discussion by AliFlux** [fastapi/fastapi#6255] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class SubType(Enum):
    thriller = "thriller"
    horror = "horror"

@router.get("/{id}-{subType}")
async def article_func_a(id: int, subType: SubType):
    return "show article with subtype"

@router.get("/{id}-{subId}")
async def article_func_b(id: int, subId: int):
    return "show article with subId"
```

## TL;DR

Is there a way to perform routing based on the data types of the path parameters? Instead of just the pattern of the path?

## Explanation

Hi,

I am designing an application which will have two different types of routes:

1. {id}-{subType}
  - Such as /1129-horror
3. {id}-{subId}
  - Such as /1129-32

When I use the code above, the first given route catches both types of URLs, and throws validation error. Because routing is based on the parameter, and doesn't care about the type of those parameters.

I am currently porting a Flask application, and it is doable in Flask by specifying types in routes such as: `/<int:id>-<string:subType>`. But apparently this isn't possible in FastAPI.

I can convert enums into saperate routes, but I have dozens of subTypes, and that won't be the most optimal solution. Especially when I am generating OpenAPI schema. 

### Operating System

Linux, Windows, macOS

### FastAPI Version

0.79.0

### Python Version

3.10.0

**Comment by JarroVGIT** [comment]
Switch the endpoints around. An integer can be casted to a string, but a string (or in your case, your enum values) cannot be casted to ints. They are evaluated in order, so when a request fits a signature, it will execute that path function. 

(note, I haven't tested it, can do that when I get home, but it should work).

**Comment by zoliknemet** [comment]
Here is an implementation with only one route and with query parameters:

```python
from enum import Enum
from typing import Optional

import uvicorn
from fastapi import FastAPI, APIRouter, Query

app = FastAPI()
router = APIRouter()


class SubType(Enum):
    thriller = "thriller"
    horror = "horror"


@router.get("/article")
async def article(
    id: int,
    subId: Optional[int] = Query(default=None),
    subType: Optional[SubType] = Query(default=None),
):
    if bool(subId) == bool(subType):
        raise Exception()  # can not query subId and subType in the same time
    if subId:
        # implementation...
        return "show article with subId"
    else:
        # implementation...
        return "show article with subtype"


app.include_router(router)

if __name__ == "__main__":
    uvicorn.run(app)

```

There are better designs and in my opinion you should avoid doing `/{something}-{somethingElse}` because route is defining the resource

**Comment by JarroVGIT** [comment, marked as answer]
My answer was incomplete. You do have to switch them around (as integers can and will be tried to be casted to strings) so it will first try to parse the integers and if that fails, it will parse it as enum (and throws an error if that fails as well). However, you need to provide the type of your path parameters like so (this is a fully working example, you can run this as-is):

```python
from fastapi import FastAPI
from enum import Enum
app = FastAPI()

class SubType(Enum):
    thriller = "thriller"
    horror = "horror"

@app.get("/{id}-{subId:int}")
async def article_func_b(id: int, subId: int):
    return "show article with subId"

@app.get("/{id}-{subType:str}")
async def article_func_a(id: int, subType: SubType):
    return "show article with subtype"

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000,  )
```

going to "/1-1" will respond with `show article with subId`, while going to "/1-horror" will result in `show article with subtype`. When providing anything that doesn't match, it will result in a validation error on the Enum. Going to "1-Foo" will result in:
```
{
    "detail": [
        {
            "loc": [
                "path",
                "subType",
            ],
            "msg": "value is not a valid enumeration member; permitted: 'thriller', 'horror'",
            "type": "type_error.enum",
            "ctx": {
                "enum_values": [
                    "thriller",
                    "horror",
                ]
            }
        }
    ]
}
```

**Reply by batyr993** [reply]
But what If I want to send a subType as a null, then how should I handle it? Currently it throws me the same error as ``` "value is not a valid enumeration member; permitted: ... ```, In my route I declared as such:
``` employee_type: models.EmployeeType = Form(None) ``` and it still gives me an error, if I pass a null value.
EDIT:
Maybe it works when you send arguments in query. But in a Form it doesn't pass.

**Comment by AliFlux** [comment]
Thanks for the working examples @zoliknemet and @JarroVGIT . These are good workarounds but will it also generate the appropriate OpenAPI schema? With enums and everything?

**Comment by JarroVGIT** [comment]
My solution is not a workaround, it is the proper implementation of your requirements. But to answer your question; yes it results in the correct schema (otherwise it would be a workaround). You can try and see for your self in the generated /docs:

<img width="408" alt="image" src="https://user-images.githubusercontent.com/13659033/179344491-48b47269-7345-44c3-a787-1c155bf2c0ad.png">
<img width="482" alt="image" src="https://user-images.githubusercontent.com/13659033/179344501-d7dcb125-c5de-4699-a544-1adbf135528c.png">

**Comment by AliFlux** [comment]
Thanks for the solution. It really works 🚀
